### PR TITLE
NRF: Implement i2c interface

### PR DESCRIPTION
**Pull request by deadprogram**
This PR adds the I2C implementation for NRF. It should work for both NRF52 and NRF51, but I have only tested with the former, so far.

### CODE REVIEW

File: src/examples/blinkm/blinkm.go [hunk 1]
```diff
@@ -20,20 +20,18 @@ func main() {
 
 	count := 0
 	for {
-		machine.I2C0.WriteTo(0x09, []byte("n"))
```
Comment (aykevl) [inline src/examples/blinkm/blinkm.go]: This change in examples/blinkm looks good, but should be done in a different PR. It appears to be unrelated to the nrf implementation.
In general, I prefer such examples in different PRs / commits, but when it is implemented for the first time (like on AVR) it's useful to have a real program to test it so that's why I was fine with it.

File: src/machine/machine_nrf.go [hunk 2]
```diff
@@ -96,3 +97,96 @@ func handleUART0() {
...
+// WriteTo writes a slice of data bytes to a peripheral with a specific address.
+func (i2c I2C) WriteTo(address uint8, data []byte) {
+	nrf.TWI0.ADDRESS = nrf.RegValue(address)
+	nrf.TWI0.TASKS_STARTTX = 1
+	for _, v := range data {
+		i2c.WriteByte(v)
+	}
+
+	// Approx. 300 us needed after ending write before sending stop.
+	time.Sleep(300 * time.Microsecond)
```
Comment (aykevl) [inline src/machine/machine_nrf.go]: That's a rather magical constant. I'm pretty sure there must be a better way than just waiting for some time (which may or may not be correct in all cases, and is probably not the most efficient way to do it).

Also, `time.Sleep()` is somewhat special. When using goroutines, it will switch to a different goroutine if one is runnable so it may not always do what you think it does.

I know that I2C is a bit strange on the nrf. I've played with it before and it was hard to get working correctly. There probably is a way with enough trial and error. For a working (but rather complex) driver implementation see https://github.com/NordicSemiconductor/nrfx/blob/master/drivers/src/nrfx_i2s.c.

File: src/machine/machine_nrf.go [hunk 3]
```diff
@@ -96,3 +97,96 @@ func handleUART0() {
...
+	nrf.TWI0.ADDRESS = nrf.RegValue(address)
+	nrf.TWI0.TASKS_STARTRX = 1
+	for i, _ := range data {
+		data[i] = i2c.ReadByte()
+	}
+
+	// Assume stop after read.
+	i2c.Stop()
+
+	return
```
Comment (aykevl) [inline src/machine/machine_nrf.go]: This return seems to be unnecessary?

File: src/examples/blinkm/blinkm.go (hunk 1 above)
Comment (deadprogram) [reply src/examples/blinkm/blinkm.go]: Actually it is needed here. The AVR is forgiving enough that the extra STOP did not affect it, but it caused problems on the nrf. Changing this works on both.

**Comment by aykevl** [comment]
> https://travis-ci.com/aykevl/tinygo/jobs/149706752#L1088

Right. That is a less obvious import cycle. This is one of the reasons why I'd like to keep the machine package free of dependencies. Either that, or some other way needs to be thought up for using stdout in the runtime.

File: src/machine/machine_nrf.go (hunk 2 above)
Comment (deadprogram) [reply src/machine/machine_nrf.go]: The code here https://github.com/aykevl/tinygo/pull/37/files#diff-194bcd9b39ab8cf699c207bc6242f403R181 is supposed to wait for the TX to complete. Which it does, but issuing a STOP immediately afterward locks up the I2C bus.

I discovered that the pause was needed there by accident due to some logging I had added. Afterward I found that the smallest pause that would work was the 300 us. No doubt that is somehow related to the bus speed, but it is not completely clear how, as of yet.

The nrfx library does not uses the direct registers, so it is of limited usefulness. Worth noting is that the TWIM interface did not work only the TWI. Not sure exactly why that might be, something to do with DMA that you would probably be better equipped to figure out. This PR uses TWI which works on both nrf52 and nrf51.

File: src/machine/machine_nrf.go (hunk 3 above)
Comment (deadprogram) [reply src/machine/machine_nrf.go]: sure.

**Comment by deadprogram** [comment]
OK so then how to handle the need to pause here without using `time`?

**Comment by deadprogram** [comment]
OK this is interesting. I just removed the extra `time.Sleep()` because not seemingly needed by itself.

I can no longer reproduce whatever conditions were requiring that extra delay. Perhaps it was the extra pull-up resistors I had previously added which I removed from my breadboard.

In any case, please review the revised PR. Thanks.

**Comment by deadprogram** [comment]
OK last change: you can also now set the SDA/SCL pins on the different buses so you can use them both at the same time, should you so require. They use the board-specific settings as a sensible default.

**Comment by deadprogram** [comment]
Tested with a slightly modified version of colorlamp to match the different pin mappings on the pca10040 vs the Arduino, and it worked perfectly.

File: src/machine/machine_nrf.go [hunk 4]
```diff
@@ -96,3 +96,109 @@ func handleUART0() {
 		nrf.UART0.EVENTS_RXDRDY = 0x0
 	}
 }
+
+// There are 2 I2C interfaces on the NRF.
+var (
+	I2C0 = I2C{Bus: 0}
+	I2C1 = I2C{Bus: 1}
```
[the remark this answers is not in the document]
Comment (deadprogram) [reply src/machine/machine_nrf.go]: Good idea. Done.

File: src/machine/i2c.go [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+const (
+	TWI_FREQ_100KHZ = 100000
+	TWI_FREQ_400KHZ = 400000
+)
+
+// I2CConfig does not do much of anything on Arduino.
+type I2CConfig struct {
+	Frequency uint32
+	SCL       uint8
+	SDA       uint8
```
[the remark this answers is not in the document]
Comment (deadprogram) [reply src/machine/i2c.go]: OK, sure. Change made.

File: src/machine/machine_nrf.go [hunk 6]
```diff
@@ -96,3 +96,109 @@ func handleUART0() {
...
+	// Default I2C bus speed is 100 kHz.
+	if config.Frequency == 0 {
+		config.Frequency = TWI_FREQ_100KHZ
+	}
+	// Default I2C pins.
+	if config.SDA == 0 {
+		config.SDA = SDA_PIN
+	}
+	if config.SCL == 0 {
+		config.SCL = SCL_PIN
```
[the remark this answers is not in the document]
Comment (deadprogram) [reply src/machine/machine_nrf.go]: Done.

File: src/examples/blinkm/blinkm.go [hunk 7]
```diff
@@ -20,20 +20,18 @@ func main() {
...
-
 		switch count {
 		case 0:
 			// Crimson
-			machine.I2C0.WriteTo(0x09, []byte{0xdc, 0x14, 0x3c})
+			machine.I2C0.WriteTo(0x09, []byte{byte('n'), 0xdc, 0x14, 0x3c})
```
[the remark this answers is not in the document]
Comment (deadprogram) [reply src/examples/blinkm/blinkm.go]: You are correct. Done.

**Comment by deadprogram** [comment]
OK updated from all feedback items. Ready for review.

**Comment by deadprogram** [comment]
Approved but not merged?

**Comment by aykevl** [comment]
I was busy with the merge, which failed at first due to mixing ADC/PWM commits.
Merged in e4fa1a8288f7f0d539cbf0df26d554cac89ab8a6.

**Comment by deadprogram** [comment]
:bowing_man: thanks.
